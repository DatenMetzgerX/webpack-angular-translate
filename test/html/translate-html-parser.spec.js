const translateDirectiveTranslationExtractor = require("../../dist/html/translate-directive-translation-extractor")
  .default;
var StatefulHtmlParser = require("../../dist/html/translate-html-parser")
  .default;
var Translation = require("../../dist/translation").default;

require("../translate-jest-matchers");

describe("StatefulHtmlParserSpecs", function() {
  "use strict";

  var loaderContext;

  beforeEach(function() {
    loaderContext = {
      registerTranslation: jest.fn(),
      emitError: jest.fn(),
      emitWarning: jest.fn(),
      resourcePath: "path/test.html",
      context: "path"
    };
  });

  describe("<translate>", function() {
    it("uses the element text as translation id", function() {
      parse("<translate>Simple</translate>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("uses the translate attribute value as id", function() {
      parse("<translate translate='simple-id'>Simple</translate>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("simple-id", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("uses the value of the translate-default as defaultText", function() {
      parse("<translate translate-default='Other default'>Simple</translate>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", "Other default", {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("only translates the attribute if the translation id is undefined", function() {
      parse("<translate translate-attr-title='Simple'></translate>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledTimes(1);
    });

    it("only translates the attribute if the translation id is empty", function() {
      parse("<translate translate-attr-title='Simple'>    </translate>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledTimes(1);
    });

    it("translates the attribute and the content of the element if translate-attr is set and the element has non empty content", function() {
      parse(
        "<translate translate-attr-title='Attribute'>Element-Text</translate>"
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Attribute", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Element-Text", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("emits an error if the content of the element is an expression", function() {
      parse("<translate>{{controller.title}}</translate>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
    });

    it("emits an error if the default text is an expression", function() {
      parse(
        "<translate translate-default='{{controller.title}}'>Simple</translate>"
      );

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
    });

    it("suppresses the error if the translation element is an expression and the element is attributed with suppress-dynamic-translation-error", function() {
      parse(
        "<translate suppress-dynamic-translation-error>{{controller.title}}</translate>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("suppresses the error if the default text is an expression and the element is attributed with suppress-dynamic-translation-error", function() {
      parse(
        "<translate translate-default='{{controller.title}}' suppress-dynamic-translation-error>simple</translate>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });
  });

  describe("<any translate>", function() {
    it("uses the value of the translate-attribute as translation id", function() {
      parse("<div translate='Simple'></div>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("uses the content as translation id if the translate-attribute has no value assigned", function() {
      parse("<div translate>Simple</div>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("uses the value of the default-text attribute as default text.", function() {
      parse("<div translate translate-default='Other default'>Simple</div>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", "Other default", {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("only translates the attribute if the element content is empty", function() {
      parse("<div translate translate-attr-title='Simple'></div>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledTimes(1);
    });

    it("only translates the attribute if the translation id is empty", function() {
      parse("<div translate translate-attr-title='Simple'></div>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Simple", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledTimes(1);
    });

    it("translates the attribute and the content of the element if translate-attr is set and the element has non empty content", function() {
      parse(
        "<div translate translate-attr-title='Attribute'>Element-Text</div>"
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Attribute", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Element-Text", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("translates the attribute and the content of the element if translate-attr is set and the translate attribute has an assigned value", function() {
      parse(
        "<div translate='Element-Text' translate-attr-title='Attribute'></div>"
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Attribute", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Element-Text", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("emits an error if the value of the translate id is an expression", function() {
      parse("<div translate='{{controller.title}}'></div>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
    });

    it("suppresses the error of a dynamic value in the translation id attribute if the element is attributed with suppress-dynamic-translation-error", function() {
      parse(
        "<div translate='{{controller.title}}' suppress-dynamic-translation-error></div>"
      );
      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("emits an error if a translation does not have a valid id", function() {
      parse("<title translate>\n     </title>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      // ensure the translation is not registered a second time because of a test if scope.text is falsy (what is the case above).
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("doesn't emit an error if a translation does not have a valid id but an attribute has been translated", function() {
      parse("<title translate translate-attr-title='test'>\n     </title>");

      expect(loaderContext.emitError).not.toHaveBeenCalled();
    });
  });

  describe("<any translate translate-attr-*>", function() {
    it("uses the value of the translate-attr-title attribute as translation-id", function() {
      parse("<div translate translate-attr-title='test'></div>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("test", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("uses the default text from the translate-default-attr-*", function() {
      parse(
        "<div translate translate-attr-title='test' translate-default-attr-title='Default Text'></div>"
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("test", "Default Text", {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("doesn't register a translation for the content of an element attributed with translate-attr", function() {
      parse(
        "<div translate translate-attr-title='test'><span>Test</span></div>"
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("test", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledTimes(1);
    });

    it("emits an error if translate-attr uses an expression", function() {
      parse(
        "<div translate translate-attr-title='{{controller.title}}'></div>"
      );

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("suppresses the error if translate-attr is used and the element is attributed with suppress-dynamic-translation-error", function() {
      parse(
        "<div translate translate-attr-title='{{controller.title}}' suppress-dynamic-translation-error></div>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });
  });

  describe("{{ any | translate }}", function() {
    it("extracts the translation id of a translate filter with a literal value", function() {
      parse("<root>{{ 'test' | translate }}</root>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("test", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 6 }
        })
      );
    });

    it("extracts the translation id if the translate filter is the first in chain and a literal value is used", function() {
      parse("<root>{{ 'test' | translate | uppercase }}</root>");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("test", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 6 }
        })
      );
    });

    it("extracts the translation id if the translate filter is used inside a text body", function() {
      parse(
        "<root>{{ ctrl.total | number:0 }} {{ 'USD' | translate }} ($)</root>"
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("USD", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 6 }
        })
      );
    });

    it("emits an error if the translate filter is being used for a dynamic value", function() {
      parse("<root>{{ controller.title | translate }}</root>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
    });

    it("emits an error if the translate filter is not the first in the filter chain", function() {
      parse("<root>{{ 'title' | uppercase | translate }}</root>");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
    });

    it("suppresses the error for a filter with a dynamic value if suppress-dynamic-translate-error is used on the parent element", function() {
      parse(
        "<span suppress-dynamic-translation-error>{{ controller.title | translate }}</span>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("suppresses the error for a filter chain where translate is not the first filter if suppress-dynamic-translate-error is used on the parent element", function() {
      parse(
        "<span suppress-dynamic-translation-error>{{ 'title' | uppercase | translate }}</span>"
      );

      expect(loaderContext.emitError).not.toHaveBeenCalled();
      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });
  });

  describe("<any attr='{{ any | translate }}'>", function() {
    it("extracts the translation from an attribute with translate filter", function() {
      parse("<img src='xy' title=\"{{ 'Waterfall' | translate }}\" />");

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Waterfall", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("extracts multiple translations from an attribute with translate filters", function() {
      parse(
        "<img src='xy' title=\"Fixed Text: {{ 'Waterfall' | translate }} and {{ 'Other' | translate }}\" />"
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Waterfall", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );

      expect(loaderContext.registerTranslation).toHaveBeenCalledWith(
        new Translation("Other", undefined, {
          resource: "path/test.html",
          loc: { line: 1, column: 0 }
        })
      );
    });

    it("emits an error if the translate filter is not the first in the filter chain", function() {
      parse("<img src='xz' alt='{{ \"title\" | uppercase | translate }}' />");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
    });

    it("emits an error if the translate filter is used for a dynamic value", function() {
      parse("<img src='xz' alt='{{ ctrl.imgAlt | translate }}' />");

      expect(loaderContext).toHaveEmittedErrorMatchingSnapshot();
    });
  });

  describe("invalid cases", function() {
    it("does not register an empty element without the translate attribute", function() {
      parse("<div></div>");

      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("does not register a translation for an element without the translate attribute", function() {
      parse("<div>Test</div>");

      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });

    it("does not register a translation for a translate-attribute if the translate directive is missing on the element", function() {
      parse("<div translate-attr-title='test'>Test</div>");

      expect(loaderContext.registerTranslation).not.toHaveBeenCalled();
    });
  });

  function parse(source) {
    var statefulParser = new StatefulHtmlParser(loaderContext, [
      translateDirectiveTranslationExtractor
    ]);
    statefulParser.parse(source);
    return statefulParser;
  }
});
