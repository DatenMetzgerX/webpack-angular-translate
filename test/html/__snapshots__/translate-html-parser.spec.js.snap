// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatefulHtmlParserSpecs <any attr='{{ any | translate }}'> emits an error if the translate filter is not the first in the filter chain 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: Another filter is used before the translate filter in the element <img src='xz' alt='{{ "title" | uppercase | translate }}'>...</img>. Add the 'suppress-dynamic-translation-error' to suppress the error (ensure that you have registered the translation manually, consider using i18n.registerTranslation).],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any attr='{{ any | translate }}'> emits an error if the translate filter is used for a dynamic value 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: A dynamic filter expression is used in the text or an attribute of the element '<img src='xz' alt='{{ ctrl.imgAlt | translate }}'>...</img>'. Add the 'suppress-dynamic-translation-error' attribute to suppress the error (ensure that you have registered the translation manually, consider using i18n.registerTranslation).],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any translate translate-attr-*> emits an error if translate-attr uses an expression 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element '<div translate='' translate-attr-title='{{controller.title}}'>...</div>' uses an angular expression as translation id ('{{controller.title}}') or as default text ('undefined'). This is not supported. Either use a string literal as translation id and default text or suppress this error by adding the 'suppress-dynamic-translation-error' attribute to this element or any of its parents.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any translate> emits an error if a translation does not have a valid id 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: the element uses the translate directive but does not specify a translation id nor has any translated attributes (translate-attr-*). Specify a translation id or remove the translate-directive.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <any translate> emits an error if the value of the translate id is an expression 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element '<div translate='{{controller.title}}'>...</div>' uses an angular expression as translation id ('{{controller.title}}') or as default text ('undefined'). This is not supported. Either use a string literal as translation id and default text or suppress this error by adding the 'suppress-dynamic-translation-error' attribute to this element or any of its parents.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <translate> emits an error if the content of the element is an expression 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element '<translate>{{controller.title}}</translate>' uses an angular expression as translation id ('{{controller.title}}') or as default text ('undefined'). This is not supported. Either use a string literal as translation id and default text or suppress this error by adding the 'suppress-dynamic-translation-error' attribute to this element or any of its parents.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs <translate> emits an error if the default text is an expression 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:0: The element '<translate translate-default='{{controller.title}}'>Simple</translate>' uses an angular expression as translation id ('Simple') or as default text ('{{controller.title}}'). This is not supported. Either use a string literal as translation id and default text or suppress this error by adding the 'suppress-dynamic-translation-error' attribute to this element or any of its parents.],
  ],
]
`;

exports[`StatefulHtmlParserSpecs {{ any | translate }} emits an error if the translate filter is being used for a dynamic value 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:6: A dynamic filter expression is used in the text or an attribute of the element '<root>{{ controller.title | translate }}</root>'. Add the 'suppress-dynamic-translation-error' attribute to suppress the error (ensure that you have registered the translation manually, consider using i18n.registerTranslation).],
  ],
]
`;

exports[`StatefulHtmlParserSpecs {{ any | translate }} emits an error if the translate filter is not the first in the filter chain 1`] = `
Array [
  Array [
    [Error: Failed to extract the angular-translate translations from 'test.html':1:6: Another filter is used before the translate filter in the element <root>{{ 'title' | uppercase | translate }}</root>. Add the 'suppress-dynamic-translation-error' to suppress the error (ensure that you have registered the translation manually, consider using i18n.registerTranslation).],
  ],
]
`;
